financial-templates-lib/clients/ExpiringMultiPartyClient.js

```diff
@@ -83,37 +83,69 @@ class ExpiringMultiPartyClient {
 
   // Returns an array of sponsor addresses.
   getAllSponsors() {
-    return this.sponsorAddresses;
+    return this.activeSponsors;
   }
 
   // Returns the last update timestamp.
   getLastUpdateTime() {
     return this.lastUpdateTimestamp;
   }
 
-  async update() {
-    // Fetch contract state variables in parallel.
-    const [collateralRequirement, liquidationLiveness, events, cumulativeFeeMultiplier] = await Promise.all([
+  async initialSetup() {
+    const [collateralRequirement, liquidationLiveness, cumulativeFeeMultiplier] = await Promise.all([
       this.emp.methods.collateralRequirement().call(),
       this.emp.methods.liquidationLiveness().call(),
-      this.emp.getPastEvents("NewSponsor", { fromBlock: 0 }),
       this.emp.methods.cumulativeFeeMultiplier().call()
     ]);
     this.collateralRequirement = this.toBN(collateralRequirement.toString());
     this.liquidationLiveness = Number(liquidationLiveness);
-    this.sponsorAddresses = [...new Set(events.map(e => e.returnValues.sponsor))];
     this.cumulativeFeeMultiplier = this.toBN(cumulativeFeeMultiplier.toString());
+  }
 
-    // Fetch sponsor position, liquidation, and current time data in parallel batches, 20 at a time, to be safe and not overload the web3 node.
-    const WEB3_CALLS_BATCH_SIZE = 20;
-    const [positions, allLiquidations, currentTime] = await Promise.all([
-      Promise.map(this.sponsorAddresses, address => this.emp.methods.positions(address).call(), {
+  async update() {
+    // If it is the first run then get contract contestants. This only needs to be called once.
+    if (!this.collateralRequirement) {
+      await this.initialSetup();
+    }
+    // Fetch contract state variables in parallel.
+    const [newSponsorEvents, endedSponsorEvents, liquidationCreatedEvents, currentTime] = await Promise.all([
+      this.emp.getPastEvents("NewSponsor", { fromBlock: 0 }),
+      this.emp.getPastEvents("EndedSponsorPosition", { fromBlock: 0 }),
+      this.emp.getPastEvents("LiquidationCreated", { fromBlock: 0 }),
+      this.emp.methods.getCurrentTime().call()
+    ]);
+
+    // Array of all sponsors, over all time. Can contain repeats for every `NewSponsor` event.
+    const newSponsorAddresses = [...newSponsorEvents.map(e => e.returnValues.sponsor)];
+
+    // Array of ended sponsors. Can contain repeats for every `EndedSponsorPosition` event.
+    const endedSponsorAddresses = [...endedSponsorEvents.map(e => e.returnValues.sponsor)];
+
```

Comment:
The logic below is a bit tricky and took me a while to get just right. The basic idea is outlined in the comments. Overall the idea is to only run the `this.emp.methods.positions` call against **active** sponsors. I use a combination of `NewSponsor` and `EndedSponsorPosition` events to build an array of _active_ sponsors.